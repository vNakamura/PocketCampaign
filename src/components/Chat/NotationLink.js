// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { rollDice } from '../../actions/chat';
import type { State, UserState } from '../../types/State';

const StyledAnchor = styled.a`
  color: ${props => props.theme.palette.text};
`;

class NotationLink extends Component {
  props: {
    notation: string,
    dispatch: Function,
    currentUser: UserState,
  }
  handleClick = (e: MouseEvent) => {
    e.preventDefault();
    this.props.dispatch(rollDice('tutorial', this.props.notation, this.props.currentUser.key));
  };
  render() {
    return (
      <StyledAnchor href="#roll" onClick={this.handleClick}>{this.props.notation}</StyledAnchor>
    );
  }
}

const mapStateToProps = (state: State) => {
  const { currentUser } = state;

  return {
    currentUser,
  };
};
export default connect(mapStateToProps)(NotationLink);
