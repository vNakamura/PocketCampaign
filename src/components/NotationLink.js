import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { setModal } from '../actions/actionCreators';
import DiceRollModal from './modals/diceRoll';

export class NotationLink extends React.Component {
  handleClick = (e) => {
    e.preventDefault();
    this.props.setModal(<DiceRollModal notation={this.props.notation} />, 'Roll Dice');
  }

  render() {
    return (
      <a
        href="#notationModal"
        onClick={this.handleClick}
      >
        {this.props.notation}
      </a>
    );
  }
}

NotationLink.propTypes = {
  notation: PropTypes.string.isRequired,
  setModal: PropTypes.func.isRequired,
};

export default connect(
  null,
  dispatch => bindActionCreators({ setModal }, dispatch),
)(NotationLink);
