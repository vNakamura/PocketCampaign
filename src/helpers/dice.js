// @flow

import React from 'react';
import droll from 'droll';
import seedrandom from 'seedrandom';
import replace from 'string-replace-to-array';

import NotationLink from '../components/Chat/NotationLink';

export const notationLimits = {
  minNumDice: 1,
  maxNumDice: 10,
  minNumSides: 2,
  maxNumSides: 100,
  minModifier: -1000,
  maxModifier: 1000,
};
export const diceNotationRegex = /\b([1-9]\d*)?d([1-9]\d*)([+-]\d+)?\b/gi;

type ResultError = {
  ok: false,
  message: string,
};
type ValidateResult = ResultError | { ok: true };
const validateRoll = (notation:string): ValidateResult => {
  const p = droll.parse(notation);
  if (!p) return { ok: false, message: 'Invalid dice notation' };
  if (p.numDice > notationLimits.maxNumDice) return { ok: false, message: 'Invalid dice notation. Too many dice!' };
  if (p.numSides < notationLimits.minNumDice) return { ok: false, message: 'Invalid dice notation. You need at least one die!' };
  if (p.numSides > notationLimits.maxNumSides) return { ok: false, message: 'Invalid dice notation. Too many sides!' };
  if (p.numSides < notationLimits.minNumSides) return { ok: false, message: 'Invalid dice notation. Too few sides!' };
  if (p.modifier > notationLimits.maxModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  if (p.modifier < notationLimits.minModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  return { ok: true };
};

export type ParseResult = ResultError | {
  ok: true,
  parsed: {
    numDice: number,
    numSides: number,
    modifier: number,
    minResult: number,
    maxResult: number,
    avgResult: number,
  }
};
export const parseRoll = (notation: string): ParseResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;
  const p = droll.parse(notation);
  return { ok: true, parsed: p };
};

export type RollResult = ResultError | {
  ok: true,
  rolls: number[],
  modifier: number,
  total: number,
};
export const roll = (notation: string, seed?: string): RollResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;

  if (seed !== undefined) seedrandom(seed, { global: true });
  const result = droll.roll(notation);
  return {
    ok: true,
    rolls: result.rolls,
    modifier: result.modifier,
    total: result.total,
  };
};

export const addLinksToText = (text: string) => {
  let i = 0;
  const result = replace(
    text,
    diceNotationRegex,
    (notation, numDice, numSides) => {
      i += 1;
      const v: ValidateResult = validateRoll(notation);
      if (v.ok && numSides > 1) return <NotationLink key={i} notation={notation} />;
      return notation;
    },
  );
  return result;
};
