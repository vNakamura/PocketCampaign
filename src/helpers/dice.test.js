import React from 'react';
import { shallow } from 'enzyme';
import { parseRoll, roll, addLinksToText } from './dice';
import NotationLink from '../components/Chat/NotationLink';

describe('helpers: dice', () => {
  describe('parseRoll', () => {
    it('should reject ivalid notations', () => {
      expect(parseRoll('d').ok).toBe(false);
      expect(parseRoll('1d').ok).toBe(false);
      expect(parseRoll('1da').ok).toBe(false);
      expect(parseRoll('1d20b').ok).toBe(false);
    });
    it('should reject notations with extreme values', () => {
      expect(parseRoll('0d6').ok).toBe(false);
      expect(parseRoll('50d6').ok).toBe(false);
      expect(parseRoll('2d1').ok).toBe(false);
      expect(parseRoll('2d500').ok).toBe(false);
      expect(parseRoll('2d20+99999').ok).toBe(false);
      expect(parseRoll('2d20-99999').ok).toBe(false);
    });
    it('should accept valid notations', () => {
      expect(parseRoll('d2').ok).toBe(true);
      expect(parseRoll('2d6').ok).toBe(true);
      expect(parseRoll('5d20+100').ok).toBe(true);
      expect(parseRoll('10d100-500').ok).toBe(true);
    });
  });
  describe('roll', () => {
    it('should validate notations before rolling', () => {
      const invalidRoll = roll('d');
      expect(invalidRoll.ok).toBe(false);
      expect(invalidRoll.message).toBeDefined();
      expect(invalidRoll.rolls).toBeUndefined();
      expect(invalidRoll.modifier).toBeUndefined();
      expect(invalidRoll.total).toBeUndefined();

      const validRoll = roll('d6');
      expect(validRoll.ok).toBe(true);
      expect(validRoll.message).toBeUndefined();
      expect(validRoll.rolls).toBeDefined();
      expect(validRoll.modifier).toBeDefined();
      expect(validRoll.total).toBeDefined();
    });

    it('should return the same result when given the same seed', () => {
      const result1 = roll('10d20', 1234);
      const result2 = roll('10d20', 1234);
      expect(result1).toMatchObject(result2);

      const result3 = roll('10d20', 1234);
      const result4 = roll('10d20', 1235);
      expect(result3).not.toMatchObject(result4);
    });

    it('should roll the exact number of dice in the notation', () => {
      const result = roll('10d10');
      expect(result.rolls).toHaveLength(10);
    });

    it('should add the given modifier to the rolls', () => {
      const notation = '10d6';
      const noModifier = roll(notation, 1);
      const withModifier = roll(`${notation}+50`, 1);
      const withNegativeModifier = roll(`${notation}-50`, 1);

      expect(noModifier.modifier).toBe(0);
      expect(withModifier.modifier).toBe(50);
      expect(withModifier.total - noModifier.total).toBe(50);
      expect(withNegativeModifier.modifier).toBe(-50);
      expect(withNegativeModifier.total - noModifier.total).toBe(-50);
    });
  });
  describe('addLinksToText', () => {
    it('should add the NotationLink component to texts with notation', () => {
      const textWithLinks = addLinksToText('Roll d20+3 for initiative.');
      expect(textWithLinks).toMatchSnapshot();
      expect(shallow(<p>{textWithLinks}</p>).find(NotationLink)).toHaveLength(1);
    });
    it('should NOT add the NotationLink component to texts when notation has too big values', () => {
      const textWithLinks = addLinksToText('Roll d200+3 for initiative.');
      expect(textWithLinks).toMatchSnapshot();
      expect(shallow(<p>{textWithLinks}</p>).find(NotationLink)).toHaveLength(0);
    });
    it('should accept multiple notations', () => {
      const textWithLinks = addLinksToText('2d10+10, 5D6, 100d20; 1d or d100.');
      expect(textWithLinks).toMatchSnapshot();
      expect(shallow(<p>{textWithLinks}</p>).find(NotationLink)).toHaveLength(3);
    });
  });
});
