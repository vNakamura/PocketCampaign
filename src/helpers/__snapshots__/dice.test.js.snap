// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers: dice addLinksToText should NOT add the NotationLink component to texts when notation has too big values 1`] = `
Array [
  "Roll ",
  "d200+3",
  " for initiative.",
]
`;

exports[`helpers: dice addLinksToText should accept multiple notations 1`] = `
Array [
  <Connect(NotationLink)
    notation="2d10+10"
/>,
  ", ",
  <Connect(NotationLink)
    notation="5D6"
/>,
  ", ",
  "100d20",
  "; 1d or ",
  <Connect(NotationLink)
    notation="d100"
/>,
  ".",
]
`;

exports[`helpers: dice addLinksToText should add the NotationLink component to texts with notation 1`] = `
Array [
  "Roll ",
  <Connect(NotationLink)
    notation="d20+3"
/>,
  " for initiative.",
]
`;
